src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -643,28 +643,23 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                 if ns != Namespace::ValueNS {
                     return None;
                 }
-                debug!("looking for variants or fields named {} for {:?}", item_name, did);
+                debug!("looking for fields named {} for {:?}", item_name, did);
                 // FIXME: this doesn't really belong in `associated_item` (maybe `variant_field` is better?)
-                // NOTE: it's different from variant_field because it resolves fields and variants,
+                // NOTE: it's different from variant_field because it only resolves struct fields,
                 // not variant fields (2 path segments, not 3).
```

Comment:
Perhaps it's worth explicitly saying that we don't need to handle variant fields in here because they can't be mistaken for associated items? At least that's the impression I got.

Reply:
Variant fields are different because they have a different syntax: fields look like `path::to::MyAdt::field`, variant fields look like `path::to::MyEnum::Variant::field`. In the first form, the type ends one path segment from the end, in the second it ends two path segments away. Ideally these *would* be combined.

Reply:
Let me give my understanding just so we're on the same page: Only associated items and struct/union fields are handled here. The reason is that they have the same syntax: `path::to::MyAdt::Assoc` and `path::to::MyAdt::field`. So it's impossible for this code to be handling a variant field since they have an extra path segment.

I think we're saying the same thing. I was just suggesting to elaborate a bit in this comment about why struct fields have to be handled at all. Currently it only says that they are and that variant fields are not; it doesn't say why struct fields are handled here.

r=me with or without this elaboration (though I'd prefer with)

Reply:
> Let me give my understanding just so we're on the same page: Only associated items and struct/union fields are handled here. The reason is that they have the same syntax: path::to::MyAdt::Assoc and path::to::MyAdt::field. So it's impossible for this code to be handling a variant field since they have an extra path segment.

Yes, that's right :)

Reply:
I just realized that I'm not sure why this code doesn't handle variants since they also look like associated items. However, the new code is strictly more correct than the previous code. I added a note in the comment that I'm unclear why variants aren't handled.